python/setuputils.py

```diff
@@ -46,7 +68,121 @@ def clean_folder(path):
+
+            raft_path = '../' + raft_path
+
+        else:
+            print("-- Using RAFT_PATH variable, RAFT found at " +
+                  str(os.environ['RAFT_PATH']))
+            raft_path = os.environ['RAFT_PATH']
+
+        try:
+            os.symlink(raft_path + 'python/raft', 'cuml/raft')
+        except FileExistsError:
+            os.remove('cuml/raft')
+            os.symlink(raft_path + 'python/raft', 'cuml/raft')
+
+    else:
+        print("-- Using already existing RAFT folder, source located at")
+        print(os.path.realpath('cuml/raft'))
+
+
+def clone_repo_if_needed(name, cpp_build_path,
+                         git_info_file='../cpp/cmake/Dependencies.cmake'):
+    if cpp_build_path:
+        cpp_build_path = '../' + cpp_build_path
+    else:
+        cpp_build_path = '../cpp/build/'
+
+    repo_cloned = get_submodule_dependency(name,
+                                           cpp_build_path=cpp_build_path,
+                                           git_info_file=git_info_file)
+
+    if repo_cloned:
+        repo_path = '_external_repositories/' + name + '/'
+    else:
+        repo_path = os.path.join(cpp_build_path, name + '/src/' + name + '/')
+
+    return repo_path, repo_cloned
+
+
+def get_submodule_dependency(repo,
+                             git_info_file='../cpp/cmake/Dependencies.cmake',
+                             cpp_build_path='../cpp/build/'):
+    """
+    Function to check if sub repositories (i.e. submodules in git terminology)
+    already exist in the libcuml build folder, otherwise will clone the
+    repos needed to build the cuML Python package.
+
+    Parameters
+    ----------
+    repo : Strings or list of Strings
+        Name of the repos to be cloned. Must match
+        the names of the cmake git clone instruction
+        `ExternalProject_Add(name`
+    git_info_file : String
+        Relative path of the location of the CMakeLists.txt (or the cmake
+        module which contains ExternalProject_Add definitions) to extract
+        the information. By default it will look in the standard location
+        `cuml_repo_root/cpp`
+    cpp_build_path : String
+        Relative location of the build folder to look if repositories
+        already exist
+
+    Returns
+    -------
+    result : boolean
+        True if repos were found in libcuml cpp build folder, False
+        if they were not found.
+    """
+
+    if isinstance(repo, str):
+        repos = [repo]
+
+    repo_info = get_repo_cmake_info(repos, git_info_file)
+
+    if os.path.exists(cpp_build_path):
+        print("-- Third party modules found succesfully in the libcuml++ "
+              "build folder.")
+
+        return False
+
+    else:
+
+        print("-- Third party repositories have not been found so they " +
+              "will be cloned. To avoid this set the environment " +
+              "variable CUML_BUILD_PATH, containing the relative " +
```

Comment:
Why does it have to be relative? Abs paths could be clearer

Reply:
Removed all relative imports, I think everything being absolute is just better for consistency, what do you think?